source/opt/folding_rules.cpp

```diff
@@ -1712,16 +1711,11 @@ bool CompositeExtractFeedingConstruct(
   // The last check it to see that the object being extracted from is the
   // correct type.
   Instruction* original_inst = def_use_mgr->GetDef(original_id);
-  analysis::TypeManager* type_mgr = context->get_type_mgr();
-  const analysis::Type* original_type =
+  uint32_t original_type_id =
       GetElementType(original_inst->type_id(), first_element_inst->begin() + 3,
-                     first_element_inst->end() - 1, type_mgr);
-
-  if (original_type == nullptr) {
-    return false;
-  }
+                     first_element_inst->end() - 1, def_use_mgr);
 
-  if (inst->type_id() != type_mgr->GetId(original_type)) {
```

Comment:
So this is the issue right? The pointer returned by the type manager can be != than the pointer used here, so the IDs won't match, even if they would use the same ID?

Could the typeManager have a `has` function which safely compares 2 types? Or shall an equal operator be implemented for `analysis::Type` so we can correctly compare them not using the ID by hand, but using a safer function?

Reply:
No the problem is a little worst than that. If the IDs did not match when they should, then we would simply miss an opportunity.

If you look at the new test case, there are two types:

```
%structA = OpTypeStruct %ulong
%structB = OpTypeStruct %ulong
```

1. The original type is structA and the result type is structB, two different type. 
2. When these two types are added to the type manager, they are considered the same type because the type manager does not look at the result id to track the type, and everything else is the exact same. 
3. When getting the id for `original_type`, the id for structB is returned. This is "correct" because `original_type` is a struct containing a single ulong. It could "correctly" return structA or structB. 
4. The two ids are equal, so the folder thinks that the original type and result type are the same. It does the optimization, even if it is incorrect.

> Could the typeManager have a has function which safely compares 2 types?

The idea for the type manager was to be able to get a type without having to create a duplicate, so you don't want to include the type id. If you already have the type id, then the type manager is not needed. If you compare the types, there is nothing to distinguish them. Even if we compared the types directly, we would have the same problems.